From bfb82878b86798de732b8263bc1f15a1c790d18d Mon Sep 17 00:00:00 2001
From: wilkie <wilkie05@gmail.com>
Date: Fri, 22 Sep 2023 13:13:12 -0700
Subject: [PATCH] Allows SELENIUM_URL to target a remote selenium instance.

This is useful for docker environments where you might spawn a docker
container for selenium via docker compose, e.g.
---
 dashboard/test/ui/features/support/connect.rb | 3 ++-
 dashboard/test/ui/runner.rb                   | 2 +-
 dashboard/test/ui/utils/selenium_browser.rb   | 8 +++++++-
 3 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/dashboard/test/ui/features/support/connect.rb b/dashboard/test/ui/features/support/connect.rb
index e6e26170fa2..18c9cd6e739 100644
--- a/dashboard/test/ui/features/support/connect.rb
+++ b/dashboard/test/ui/features/support/connect.rb
@@ -60,7 +60,8 @@ def get_browser(test_run_name)
   browser = nil
   if ENV['TEST_LOCAL'] == 'true'
     headless = ENV['TEST_LOCAL_HEADLESS'] == 'true'
-    browser = SeleniumBrowser.local(headless, ENV['BROWSER_CONFIG'])
+
+    browser = SeleniumBrowser.local(headless, ENV['BROWSER_CONFIG'], ENV['SELENIUM_URL'])
   else
     browser = Retryable.retryable(tries: MAX_CONNECT_RETRIES) do
       saucelabs_browser(test_run_name)
diff --git a/dashboard/test/ui/runner.rb b/dashboard/test/ui/runner.rb
index 5deea2bcd82..637853879b6 100755
--- a/dashboard/test/ui/runner.rb
+++ b/dashboard/test/ui/runner.rb
@@ -138,7 +138,7 @@ def parse_options
       opts.on("-f", "--feature Feature", Array, "Single feature or comma separated list of features to run") do |f|
         options.features = f
       end
-      opts.on("-l", "--local", "Use local domains. Also use local webdriver (not Saucelabs) unless -c is specified.") do
+      opts.on("-l", "--local", "Use local domains. Also use local webdriver (not Saucelabs) unless -c or the SELENIUM_URL environment variable is specified.") do
         options.local = 'true'
         options.pegasus_domain = 'localhost.code.org:3000'
         options.dashboard_domain = 'localhost-studio.code.org:3000'
diff --git a/dashboard/test/ui/utils/selenium_browser.rb b/dashboard/test/ui/utils/selenium_browser.rb
index add4047ea9b..d8adaea7b7f 100644
--- a/dashboard/test/ui/utils/selenium_browser.rb
+++ b/dashboard/test/ui/utils/selenium_browser.rb
@@ -2,7 +2,7 @@ require 'selenium/webdriver'
 require 'webdrivers'
 
 module SeleniumBrowser
-  def self.local(headless = true, browser = :chrome)
+  def self.local(headless = true, browser = :chrome, selenium_url = nil)
     browser = browser.to_sym
     options = {}
     case browser
@@ -15,6 +15,12 @@ module SeleniumBrowser
       options[:options].headless! if headless
       options[:options].add_argument('window-size=1280,1024')
     end
+
+    if selenium_url
+      browser = :remote
+      options[:url] = selenium_url
+    end
+
     Selenium::WebDriver.for browser, options
   end
 
-- 
2.30.2

